feat(openadapt.privacy.providers): add private_ai as a scrubbing provider

KrishPatel13 opened the pull request:
<!-- Thank you for submitting a pull request! To ensure a prompt review of your changes, please provide the following information. -->

**What kind of change does this PR introduce?**
Add PrivateAI as Scrubbing Provider.
<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Summary**
Motivation: https://github.com/OpenAdaptAI/OpenAdapt/pull/418#issuecomment-1692155418
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? Try to link to an open issue. -->

**Checklist**
* [x] My code follows the style guidelines of OpenAdapt
* [x] I have performed a self-review of my code
* [x] If applicable, I have added tests to prove my fix is functional/effective
* [x] I have linted my code locally prior to submission
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation (e.g. README.md, requirements.txt)
* [x] New and existing unit tests pass locally with my changes

**How can your code be run and tested?**

<!-- See the README.md for examples. Include test output. -->


**Other information**
<!-- Delete this subheading if no additional context is needed. -->

KrishPatel13 commented:
Now, add tests for text, pdf and image scrubbing using private ai

KrishPatel13 commented:
Now, if there is some issue with PRIVATE_AI_API_KEY (not valid or not present) then all then tests in test_private_ai_scrub.py get skipped. 

See how it will look:

```
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\my_forked\OpenAdapt> pytest .\tests\openadapt\privacy\providers\test_private_ai_scrub.py
=============================================================================================== test session starts ================================================================================================
platform win32 -- Python 3.10.11, pytest-7.1.3, pluggy-1.3.0
rootdir: P:\OpenAdapt AI - MLDS AI\cloned_repo\my_forked\OpenAdapt, configfile: pyproject.toml
plugins: anyio-3.7.1
collected 15 items

tests\openadapt\privacy\providers\test_private_ai_scrub.py sssssssssssssss                                                                                                                                    [100%]

=============================================================================================== 15 skipped in 0.44s ================================================================================================ 
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\my_forked\OpenAdapt> 
```

KrishPatel13 commented:
Pytest:

```
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\my_forked\OpenAdapt> pytest
=========================================================================== test session starts ============================================================================  
platform win32 -- Python 3.10.11, pytest-7.1.3, pluggy-1.3.0
rootdir: P:\OpenAdapt AI - MLDS AI\cloned_repo\my_forked\OpenAdapt, configfile: pyproject.toml
plugins: anyio-3.7.1
collected 60 items

tests\openadapt\test_crop.py .                                                                                                                                        [  1%]  
tests\openadapt\test_events.py .......                                                                                                                                [ 13%]  
tests\openadapt\test_summary.py ..                                                                                                                                    [ 16%]  
tests\openadapt\privacy\test_providers.py .                                                                                                                           [ 18%]  
tests\openadapt\privacy\providers\test_comprehend_scrub.py ...............                                                                                            [ 43%]  
tests\openadapt\privacy\providers\test_presidio_scrub.py .................                                                                                            [ 71%]  
tests\openadapt\privacy\providers\test_private_ai_scrub.py .................                                                                                           [100%]

============================================================================= warnings summary ============================================================================== 
tests/openadapt/privacy/providers/test_presidio_scrub.py::test_scrub_image
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-vs6jmx3S-py3.10\lib\site-packages\pytesseract\pytesseract.py:382: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    return LooseVersion(

tests/openadapt/privacy/providers/test_presidio_scrub.py::test_scrub_image
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-vs6jmx3S-py3.10\lib\site-packages\setuptools\_distutils\version.py:345: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    other = LooseVersion(other)

tests/openadapt/privacy/providers/test_private_ai_scrub.py::test_image_redaction
tests/openadapt/privacy/providers/test_private_ai_scrub.py::test_image_redaction
tests/openadapt/privacy/providers/test_private_ai_scrub.py::test_image_redaction
tests/openadapt/privacy/providers/test_private_ai_scrub.py::test_image_redaction
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-vs6jmx3S-py3.10\lib\site-packages\easyocr\utils.py:576: DeprecationWarning: ANTIALIAS is deprecated and will be removed in Pillow 10 (2023-07-01). Use LANCZOS or Resampling.LANCZOS instead.
    img = cv2.resize(img,(int(model_height*ratio),model_height),interpolation=Image.ANTIALIAS)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================ 60 passed, 6 warnings in 200.61s (0:03:20) ================================================================= 
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\my_forked\OpenAdapt>
```

KrishPatel13 commented:
@abrichr Ready for review!

KrishPatel13 commented on assets/sample_llc_1.pdf:
remove files from assets/ and move them to tests/assets

KrishPatel13 commented on pyproject.toml:
```diff
@@ -86,6 +86,7 @@ click = "^8.1.6"
 spacy-transformers = "^1.2.5"
 boto3 = "^1.28.30"
 botocore = "^1.31.30"
+easyocr = "^1.7.0"
```
add `easyocr` for extracting text from image for `test_imafe_redaction` in `test_privateai_scrub.py`

KrishPatel13 commented on tests/openadapt/privacy/providers/test_comprehend_scrub.py:
```diff
@@ -17,22 +17,6 @@
     pytestmark = pytest.mark.skip(reason=msg)
 
 
-def _hex_to_rgb(hex_color: int) -> tuple[int, int, int]:
```
remove unused function

KrishPatel13 commented on tests/openadapt/privacy/providers/test_private_ai_scrub.py:
```diff
@@ -0,0 +1,218 @@
+    if type(response) is dict and "details" in response.keys():
+        raise ValueError(response.get("detail"))
+
+    assert response.get("entities_present") is False
```
Check if there are any PII/PHI entities in redacted pdf using the PrivateAI API

KrishPatel13 commented on tests/openadapt/privacy/providers/test_private_ai_scrub.py:
```diff
@@ -0,0 +1,218 @@
+    redact_text = scrub.scrub_text(extract_redacted_text)
+
+    assert redact_text == extract_redacted_text
+
```
Use easy ocr to extract text from redacted image and check if same after scrubbing/redacting via scrub_text function of PrivateAIScrubbingProvider

KrishPatel13 reviewed (commented):
Self-reviewed!

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,184 @@
+        """
+        url = "https://api.private-ai.com/deid/v3/process/files/base64"
+        file_dir = "assets/"
+        file_name = "temp_image_to_scrub.png"
```
If these are constants, can you please name them as such? e.g. TEMP_DIR_NAME, TEMP_FILE_NAME

KrishPatel13 replied:
addressed in: [6bc38d5](https://github.com/OpenAdaptAI/OpenAdapt/pull/486/commits/6bc38d5a3fec74f0c64683258cb814aeac0db4f7)

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,184 @@
+        temp_image_path = os.path.join(file_dir, file_name)
+        image.save(temp_image_path)
+
+        file_type = "image/png"
```
Please name as constant, i.e. `CONTENT_TYPE`

KrishPatel13 replied:
addressed in: [6bc38d5](https://github.com/OpenAdaptAI/OpenAdapt/pull/486/commits/6bc38d5a3fec74f0c64683258cb814aeac0db4f7)

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,184 @@
+
+        response = requests.post(url, json=payload, headers=headers)
+        if response is None:
+            raise ValueError("Private AI request returned None")
```
I think you can do something like `response.raise_for_status` or similar that would implement this

KrishPatel13 replied:
addressed in commit

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,184 @@
+            raise ValueError("Private AI request returned None")
+        response = response.json()
+        logger.debug(response)
+        if type(response) is dict and "detail" in response.keys():
```
Isn't `type(response.json())` always a dict? If so, can you please remove the unnecessary check?

abrichr replied:
Please add a comment explaining when the response is list and when it's a dict

KrishPatel13 replied:
Addressed in: [7a68b77](https://github.com/OpenAdaptAI/OpenAdapt/pull/486/commits/7a68b77c8926486d9e4602d3361e09a721b8f0cb)

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,184 @@
+        file_type = "image/png"
+
+        # Read from file
+        with open(temp_image_path, "rb") as b64_file:
```
Can we please use `io.BytesIO` instead of writing to a file? https://docs.python.org/3/library/io.html

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,184 @@
+
+        # save file as "temp_image_to_scrub.png in assets/
+        temp_image_path = os.path.join(file_dir, file_name)
+        image.save(temp_image_path)
```
Can we please use io.BytesIO instead of writing to a file? https://docs.python.org/3/library/io.html

KrishPatel13 replied:
Addressed in: e1f4c19416dde108643be94a8f8972cb7fbd80f5

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,184 @@
+
+        response = requests.post(url, json=payload, headers=headers)
+        if response is None:
+            raise ValueError("Private AI request returned None")
```
`raise_for_status`?

KrishPatel13 replied:
On researching, I found that, PrivateAI never returns a HTTP error,
For all 4x and 5xx HTTP errors, it returns a dict of this format:
{ "detail": "<message due to error>" }

KrishPatel13 replied:
Addressed in: f486cf6e3d808bd0ffd68fc97c17213226f7be6a

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,184 @@
+        response = response.json()
+        if type(response) is dict and "details" in response.keys():
+            raise ValueError(response.get("detail"))
+        logger.debug(response.get("entities"))
```
Please use f-strings to indicate what is being logged, i.e.

```
        logger.debug(f"{response.get('entities')=}")
```

KrishPatel13 replied:
Addressed in: a37ee99ecc6ce1bed52b3cc3918c93495c06b785

abrichr commented on tests/openadapt/privacy/test_providers.py:
```diff
@@ -22,8 +23,10 @@ def test_scrubbing_provider_factory() -> None:
+            isinstance(provider, PresidioScrubbingProvider)
+            or isinstance(provider, ComprehendScrubbingProvider)
+            or isinstance(provider, PrivateAIScrubbingProvider)
         )
```
This seems unnecessary, since we already know the return value of get_for_modality is a list of providers.

KrishPatel13 replied:
Addressed in b54bc6638be20a9ef9767c317f91021d842dd091

abrichr commented on tests/openadapt/privacy/providers/test_private_ai_scrub.py:
```diff
@@ -0,0 +1,218 @@
+        raise ValueError("Private AI request returned None")
+    response = response.json()
+    if type(response) is dict and "details" in response.keys():
+        raise ValueError(response.get("detail"))
```
`.get` is unnecessary if we're already checking that `'details'` is in `responsel.keys()`.

Also, you should only need to do '"details" in response`

KrishPatel13 replied:
Addressed in: [e743b6c](https://github.com/OpenAdaptAI/OpenAdapt/pull/486/commits/e743b6ce460b5213d3b4ce06717cbe9236a09aef)

abrichr commented on tests/openadapt/privacy/providers/test_private_ai_scrub.py:
```diff
@@ -0,0 +1,218 @@
+    pdf_path = "tests/assets/sample_llc_1.pdf"
+    redacted_pdf_path = scrub.scrub_pdf(pdf_path)
+
+    url = "https://api.private-ai.com/deid/v3/process/files/base64"
```
Please define as a constant

KrishPatel13 replied:
addressed in: 74399ce3d7eb0ae8d20f01cede1a3906a7534703

abrichr commented on tests/openadapt/privacy/providers/test_private_ai_scrub.py:
```diff
@@ -0,0 +1,218 @@
+
+    url = "https://api.private-ai.com/deid/v3/process/files/base64"
+
+    file_type = "application/pdf"
```
`CONTENT_TYPE`

KrishPatel13 replied:
Addressed in: 74399ce3d7eb0ae8d20f01cede1a3906a7534703

KrishPatel13 commented:
@abrichr Addressed all of your comments!

Ready for merging

KrishPatel13 commented:
https://github.com/OpenAdaptAI/OpenAdapt/pull/418#issuecomment-1701986460

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,191 @@
+        redact_file_path = os.path.join(FILES_DIR, f"redacted-{TEMP_IMAGEFILE_NAME}")
+
+        # Write to file
+        with open(redact_file_path, "wb") as redacted_file:
```
What do you think about renaming `redact_file_path` to `redacted_file_path` to be consistent with `redacted_file_data`?

Also, what do you think about using BytesIO here as well, and elsewhere a file is written in this PR?

KrishPatel13 replied:
>> Also, what do you think about using BytesIO here as well, and elsewhere a file is written in this PR?

Addressed in: [2f92260](https://github.com/OpenAdaptAI/OpenAdapt/pull/486/commits/2f922602e139e32bead068caf961238b23fb3b6a)

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,191 @@
+        redacted_file_path = path_to_pdf.split(".")[0] + "_redacted.pdf"
+
+        # Write to file
+        with open(redacted_file_path, "wb") as redacted_file:
```
What do you think about using `BytesIO` here?

KrishPatel13 replied:
Addressed in :[2f92260](https://github.com/OpenAdaptAI/OpenAdapt/pull/486/commits/2f922602e139e32bead068caf961238b23fb3b6a)

KrishPatel13 commented:
@abrichr I have addressed your comments again, to use BytesIO wherever possible.

Here is how pytest looks like after all:

```
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\my_forked\OpenAdapt> pytest
==================================================== test session starts =====================================================
platform win32 -- Python 3.10.11, pytest-7.1.3, pluggy-1.3.0
rootdir: P:\OpenAdapt AI - MLDS AI\cloned_repo\my_forked\OpenAdapt, configfile: pyproject.toml
plugins: anyio-3.7.1
collected 60 items

tests\openadapt\test_crop.py .                                                                                          [  1%]
tests\openadapt\test_events.py .......                                                                                  [ 13%]
tests\openadapt\test_summary.py ..                                                                                      [ 16%]
tests\openadapt\privacy\test_providers.py .                                                                             [ 18%]
tests\openadapt\privacy\providers\test_comprehend_scrub.py ...............                                              [ 43%]
tests\openadapt\privacy\providers\test_presidio_scrub.py .................                                                                      [ 71%]
tests\openadapt\privacy\providers\test_private_ai_scrub.py .................                                                                    [100%]

================================================================== warnings summary ==================================================================
tests/openadapt/privacy/providers/test_presidio_scrub.py::test_scrub_image
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-vs6jmx3S-py3.10\lib\site-packages\pytesseract\pytesseract.py:382: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    return LooseVersion(

tests/openadapt/privacy/providers/test_presidio_scrub.py::test_scrub_image
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-vs6jmx3S-py3.10\lib\site-packages\setuptools\_distutils\version.py:345: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    other = LooseVersion(other)

tests/openadapt/privacy/providers/test_private_ai_scrub.py::test_image_redaction
tests/openadapt/privacy/providers/test_private_ai_scrub.py::test_image_redaction
tests/openadapt/privacy/providers/test_private_ai_scrub.py::test_image_redaction
tests/openadapt/privacy/providers/test_private_ai_scrub.py::test_image_redaction
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-vs6jmx3S-py3.10\lib\site-packages\easyocr\utils.py:576: DeprecationWarning: ANTIALIAS is deprecated and will be removed in Pillow 10 (2023-07-01). Use LANCZOS or Resampling.LANCZOS instead.
    img = cv2.resize(img,(int(model_height*ratio),model_height),interpolation=Image.ANTIALIAS)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================================================== 60 passed, 6 warnings in 280.52s (0:04:40) ===================================================== 
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\my_forked\OpenAdapt> 
```

Ready for merging!

abrichr reviewed (approved):
LGTM!

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,198 @@
+        if type(data) is dict and "detail" in data:
+            raise ValueError(data.get("detail"))
+
+        redacted_text = data[0].get("processed_text")
```
Looks like this is raising an error in https://github.com/OpenAdaptAI/OpenAdapt/actions/runs/6609799609/job/17950542911?pr=486 :

```
==================================== ERRORS ====================================
_ ERROR collecting tests/openadapt/privacy/providers/test_private_ai_scrub.py __
tests/openadapt/privacy/providers/test_private_ai_scrub.py:18: in <module>
    scrub.scrub_text("hello Bob smith")
openadapt/privacy/providers/private_ai.py:73: in scrub_text
    redacted_text = data[0].get("processed_text")
E   KeyError: 0
------------------------------- Captured stderr --------------------------------
2023-10-23 07:05:11.403 | DEBUG    | openadapt.privacy.providers.private_ai:scrub_text:63 - data={'statusCode': 401, 'message': 'Access denied due to invalid subscription key. Make sure to provide a valid key for an active subscription.'}
=============================== warnings summary ===============================
../../../../Library/Caches/pypoetry/virtualenvs/openadapt-NqG5nKAR-py3.10/lib/python3.10/site-packages/onnxruntime/capi/_pybind_state.py:28
  /Users/runner/Library/Caches/pypoetry/virtualenvs/openadapt-NqG5nKAR-py3.10/lib/python3.10/site-packages/onnxruntime/capi/_pybind_state.py:28: DeprecationWarning: invalid escape sequence '\S'
    "(other than %SystemRoot%\System32), "

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
ERROR tests/openadapt/privacy/providers/test_private_ai_scrub.py - KeyError: 0
!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
========================= 1 warning, 1 error in 18.57s =========================
```
It looks like a few things need to happen here:
1. Add a PrivateAI API key to the repo's environment variables. @KrishPatel13 can you please remind me whether we registered a key for OpenAdapt?
2. Fix the code here to handle this scenario
3. Modify the test not to run if the API key is not defined. It looks like this was implemented (https://github.com/OpenAdaptAI/OpenAdapt/pull/486/files#diff-2b5c61fcc6e55fe8105e356530deb8560174d69b8c1066cf9d62676f801aee0bR22) but is not functioning correctly.

abrichr commented on openadapt/privacy/providers/private_ai.py:
```diff
@@ -0,0 +1,198 @@
+        }
+
+        response = requests.post(TEXT_URL, json=payload, headers=headers)
+
```
```suggestion
        response.raise_for_status()
```

abrichr commented on tests/openadapt/privacy/providers/test_private_ai_scrub.py:
```diff
@@ -0,0 +1,232 @@
+
+try:
+    scrub.scrub_text("hello Bob smith")
+except ValueError:
```
```suggestion
except ValueError, requests.exceptions.HTTPError:
```

abrichr commented on tests/openadapt/privacy/providers/test_private_ai_scrub.py:
```diff
@@ -0,0 +1,232 @@
+
+try:
+    scrub.scrub_text("hello Bob smith")
+except ValueError, requests.exceptions.HTTPError:
```
```suggestion
except (ValueError, requests.exceptions.HTTPError):
```
